demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c

```diff
@@ -37,72 +44,206 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
+
+    /* If BFB2 (Boot From Bank 2) is set, erase bank 1, otherwise erase bank 2 */
+    xEraseInitStruct.Banks =
+        ( ( xOptionBytes.USERConfig & OB_BFB2_ENABLE ) == OB_BFB2_ENABLE )
+        ? FLASH_BANK_1
+        : FLASH_BANK_2;
+    xEraseInitStruct.TypeErase = FLASH_TYPEERASE_MASSERASE;
+
+    HAL_FLASH_Unlock();
+
+    /* Erase non-boot bank */
+    if( HAL_FLASHEx_Erase( &xEraseInitStruct, &ulPageError ) != HAL_OK )
+    {
+        /* Error occurred during page erase. */
+        AZLogError( ( "Error erasing flash bank\r\n" ) );
+        xResult = eAzureIoTErrorFailed;
+    }
 
-    LogInfo( ( "AzureIoTPlatform_Init()\r\n" ) );
+    HAL_FLASH_Lock();
 
-    return eAzureIoTSuccess;
+    return xResult;
 }
 
 AzureIoTResult_t AzureIoTPlatform_WriteBlock( AzureADUImage_t * const pxAduImage,
                                               uint32_t ulOffset,
                                               uint8_t * const pData,
                                               uint32_t ulBlockSize )
 {
-    (void)pxAduImage;
-    (void)ulOffset;
-    (void)pData;
-    (void)ulBlockSize;
+    uint8_t * pucNextWriteAddr = pxAduImage->xUpdatePartition + ulOffset;
+    uint8_t * pucNextReadAddr = pData;
+    uint8_t * pucLastSectionAddr = pxAduImage->xUpdatePartition + ulOffset + ulBlockSize - azureiotflashL475_FLASH_ROW_SIZE;
+    AzureIoTResult_t xResult = eAzureIoTSuccess;
 
-    // TODO: Fill in to write pData of size ulBlockSize to ulOffset in memory.
+    HAL_FLASH_Unlock();
 
-    LogInfo( ( "AzureIoTPlatform_WriteBlock(): offset %i\r\n", ulOffset ) );
+    /* write sections 1...n-1 of all blocks */
+    while( pucNextWriteAddr < pucLastSectionAddr )
+    {
+        if( HAL_FLASH_Program( FLASH_TYPEPROGRAM_FAST, ( uint32_t ) pucNextWriteAddr, ( uint32_t ) pucNextReadAddr ) != HAL_OK )
+        {
+            /* Error occurred while writing data in Flash memory */
+            HAL_FLASH_Lock();
+            return eAzureIoTErrorFailed;
+        }
+
+        pucNextWriteAddr += azureiotflashL475_FLASH_ROW_SIZE;
+        pucNextReadAddr += azureiotflashL475_FLASH_ROW_SIZE;
+    }
 
-    return eAzureIoTSuccess;
+    /** if last block and last section of that block, use
+     * FLASH_TYPEPROGRAM_FAST_AND_LAST
+     */
+    if( pxAduImage->ulImageFileSize - ulOffset <= ulBlockSize )
+    {
+        if( HAL_FLASH_Program( FLASH_TYPEPROGRAM_FAST_AND_LAST, ( uint32_t ) pucNextWriteAddr, ( uint32_t ) pucNextReadAddr ) != HAL_OK )
+        {
+            /* Error occurred while writing data in Flash memory */
+            xResult = eAzureIoTErrorFailed;
+        }
+    }
+    else
+    {
+        /* write last section of blocks 1...n-1 normally */
+        if( HAL_FLASH_Program( FLASH_TYPEPROGRAM_FAST, ( uint32_t ) pucNextWriteAddr, ( uint32_t ) pucNextReadAddr ) != HAL_OK )
+        {
+            /* Error occurred while writing data in Flash memory */
+            xResult = eAzureIoTErrorFailed;
+        }
+    }
```

Comment:
I can't add it as a suggestion but from line 103-135, would this cover all the scenarios:

```c
    while( pucNextWriteAddr < pucLastSectionAddr )
    {
        if( HAL_FLASH_Program( pxAduImage->ulImageFileSize - ulOffset <= ulBlockSize ? 
                FLASH_TYPEPROGRAM_FAST_AND_LAST : FLASH_TYPEPROGRAM_FAST, ( uint32_t ) pucNextWriteAddr, ( uint32_t ) pucNextReadAddr ) != HAL_OK )
        {
            /* Error occurred while writing data in Flash memory */
            HAL_FLASH_Lock();
            return eAzureIoTErrorFailed;
        }

        pucNextWriteAddr += azureiotflashL475_FLASH_ROW_SIZE;
        pucNextReadAddr += azureiotflashL475_FLASH_ROW_SIZE;
    }
```
(using a ternary)

Reply:
This would work with some modification (below) - I originally didn't do it this way to cut down on the repeated if statement checking with calculations within the loop, but if we check whether it's the last block first, that should minimize the calculation of whether it's the last section in the block: `pucNextWriteAddr > pucBlockEndAddr - azureiotflashL475_FLASH_ROW_SIZE`. And I can move some of these out of the while loop to be pre-calculated. 

```
uint8_t * pucBlockEndAddr = pxAduImage->xUpdatePartition + ulOffset + ulBlockSize;
while( pucNextWriteAddr < pucBlockEndAddr )
    {
        if( HAL_FLASH_Program( (pxAduImage->ulImageFileSize - ulOffset <= ulBlockSize) && (pucNextWriteAddr > pucBlockEndAddr - azureiotflashL475_FLASH_ROW_SIZE) ? 
                FLASH_TYPEPROGRAM_FAST_AND_LAST : FLASH_TYPEPROGRAM_FAST, ( uint32_t ) pucNextWriteAddr, ( uint32_t ) pucNextReadAddr ) != HAL_OK )
        {
            /* Error occurred while writing data in Flash memory */
            HAL_FLASH_Lock();
            return eAzureIoTErrorFailed;
        }

        pucNextWriteAddr += azureiotflashL475_FLASH_ROW_SIZE;
        pucNextReadAddr += azureiotflashL475_FLASH_ROW_SIZE;
    }